src/main/cpp/AutonomousModes/AutonMotionProfile.cpp

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2017-2020 FRC Team 3512. All Rights Reserved.
+
+//=============================================================================
+// File Name: AutonMotionProfile.cpp
+// Description: Drives set distance with motion profiles
+// Author: FRC Team 3512, Spartatroniks
+//=============================================================================
+
+#include "Robot.hpp"
+
+enum class State { kInit, kMotion, kIdle };
+static State state;
+
+void Robot::AutonMotionProfileInit() { state = State::kInit; }
+
+void Robot::AutonMotionProfile() {
+    switch (state) {
+        case State::kInit: {
+            // Reload drive PID constants
+            robotDrive.reloadPID();
+            robotDrive.resetEncoders();
+            claw.SetAngle(88.f);
+
+            state = State::kInit;
+            break;
+        }
+
+        case State::kMotion: {
+            // Move robot 360 cm forward
+            m_profilePID.SetGoal(360_cm);
+            double setpoint = 0.0;
+
+            while (IsAutonomous() && IsEnabled()) {
+                setpoint = m_profilePID.Calculate(units::meter_t(setpoint));
+                robotDrive.setLeftSetpoint(setpoint);
+                robotDrive.setRightSetpoint(setpoint);
+
+                claw.Update();
+
+                frc2::Wait(0.01_s);
+
+                // Stop moving
+                robotDrive.setLeftManual(0.f);
+                robotDrive.setRightManual(0.f);
+
+                // Start shooting
+                claw.Shoot();
+
+                while (IsAutonomous() && IsEnabled() && claw.IsShooting()) {
+                    claw.Update();
+
+                    frc2::Wait(0.01_s);
+                }
+                state = State::kIdle;
+                break;
+            }
+        }
```

Comment:
Blocking code is bad. If the code needs to wait until a function returns true, make that another state in the state machine.